### PR TITLE
Add string.Format / StringBuilder.AppendFormat CompositeFormat analyzer

### PR DESCRIPTION
**Pull request by stephentoub**
Contributes to https://github.com/dotnet/runtime/issues/85525
This PR provides an analyzer. It still warrants a fixer.

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCompositeFormat.cs [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    /// arguments.
+    /// </remarks>
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class UseCompositeFormatAnalyzer : DiagnosticAnalyzer
+    {
+        internal static readonly DiagnosticDescriptor UseCompositeFormatRule = DiagnosticDescriptorHelper.Create("CA1862",
+            CreateLocalizableResourceString(nameof(UseCompositeFormatTitle)),
+            CreateLocalizableResourceString(nameof(UseCompositeFormatMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
```
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCompositeFormat.cs:31]: Isn't the API review approved it with Hidden level?
```suggestion
            RuleLevel.IdeHidden_BulkConfigurable,
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCompositeFormat.cs [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+                    !compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIFormatProvider, out INamedTypeSymbol? formatProviderType) ||
+                    !compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemTextStringBuilder, out INamedTypeSymbol? stringBuilderType))
+                {
+                    return;
+                }
+
+                // Process all calls to string.Format, assuming we can find all the members we'd use as replacements.
+                IMethodSymbol[] formatCompositeMethods = stringType.GetMembers("Format").OfType<IMethodSymbol>()
+                    .Where(m => m.IsStatic &&
+                                m.Parameters.Length >= 3 &&
```
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCompositeFormat.cs:60]: Why `>= 3`? The overload with 2 params like below could not covered?     
[`public static string Format([StringSyntax(StringSyntaxAttribute.CompositeFormat)] string format, object? arg0)`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs#L422)

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCompositeFormat.cs [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+                                SymbolEqualityComparer.Default.Equals(m.Parameters[1].Type, compositeFormatType)).ToArray();
+                if (HasAllCompositeFormatMethods(appendFormatCompositeMethods))
+                {
+                    compilationContext.RegisterOperationAction(
+                        CreateAnalysisAction(isStatic: false, stringBuilderType, "AppendFormat", formatProviderType), OperationKind.Invocation);
+                }
+            });
+        }
+
+        /// <summary>Creates a delegate to register with RegisterOperationAction and that flags all of the relevate format string parameters that warrant replacing.</summary>
```
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCompositeFormat.cs:83]: NIT:
```suggestion
        /// <summary>Creates a delegate to register with RegisterOperationAction and that flags all of the relevant format string parameters that warrant replacing.</summary>
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCompositeFormat.cs [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+                        CreateAnalysisAction(isStatic: true, stringType, "Format", formatProviderType), OperationKind.Invocation);
+                }
+
+                // Process all calls to StringBuilder.AppendFormat, assuming we can find all the members we'd use as replacements.
+                IMethodSymbol[] appendFormatCompositeMethods = stringBuilderType.GetMembers("AppendFormat").OfType<IMethodSymbol>()
+                    .Where(m => !m.IsStatic &&
+                                m.Parameters.Length >= 3 &&
+                                SymbolEqualityComparer.Default.Equals(m.Parameters[0].Type, formatProviderType) &&
+                                SymbolEqualityComparer.Default.Equals(m.Parameters[1].Type, compositeFormatType)).ToArray();
+                if (HasAllCompositeFormatMethods(appendFormatCompositeMethods))
```
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCompositeFormat.cs:75]: Just curious why need to check if all Format methods ...

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCompositeFormat.cs (hunk 2 above)
Comment (buyaa-n) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCompositeFormat.cs:60]: Ah never mind, its for overloads having `CompositeFormat format` parameter

**Comment by buyaa-n** [comment]
JFYI: got a conflicts

**Comment by stephentoub** [comment]
@buyaa-n, feedback and conflicts addressed.

**Review by buyaa-n — APPROVED** [review]
Build still failing because of generated files, seems need to run `msbuild /t:pack` once more:
```
Missing entry in D:\a\_work\1\s\src\NetAnalyzers\RulesMissingDocumentation.md
  CA1863 | <[https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1863>](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1863%3E) | Use 'CompositeFormat' |
  One or more auto-generated documentation files were either edited manually, or not updated. Please revert changes made to the following files (if manually edited) and run `msbuild /t:pack` at the root of the repo to automatically update them:
```
Otherwise LGTM, thank you!

**Pull request merged**